File: src/main/scala/com/juanoff/scala/types/StringType.scala
package com.juanoff.scala.types

//import com.juanoff.types.{Comparator, UserType}

import java.io.{BufferedReader, InputStreamReader}

class StringType extends UserType {
  override def typeName(): String = "String"

  override def create(): Any = ""

  override def clone(obj: Any): Any = {
    obj match {
      case obj: String => obj
      case _ => throw new IllegalArgumentException(s"Expected ${typeName()}")
    }
  }

  override def readValue(in: InputStreamReader): Any = {
    val reader = new BufferedReader(in)
    reader.readLine()
  }

  override def parseValue(ss: String): Any = {
    if (ss == null || ss.trim.isEmpty) {
      throw new IllegalArgumentException("Enter value")
    }
    ss.trim
  }

  override def getTypeComparator: Comparator = {
    (o1: Any, o2: Any) => o1.asInstanceOf[String].compareTo(o2.asInstanceOf[String])
  }

  override def serialize(obj: Any): String = obj.asInstanceOf[String]

  override def deserialize(s: String): Any = s
}

File: src/main/java/com/juanoff/types/StringType.java
package com.juanoff.types;

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class StringType implements UserType {

    @Override
    public String typeName() {
        return UserTypeName.STRING.getTypeName();
    }

    @Override
    public Object create() {
        return "";
    }

    @Override
    public Object clone(Object obj) {
        return obj;
    }

    @Override
    public Object readValue(InputStreamReader in) throws Exception {
        BufferedReader reader = new BufferedReader(in);
        return reader.readLine();
    }

    @Override
    public Object parseValue(String ss) {
        if (ss == null || ss.isBlank()) {
            throw new IllegalArgumentException("Enter value");
        }
        return ss.trim();
    }

    @Override
    public Comparator getTypeComparator() {
        return (o1, o2) -> ((String) o1).compareTo((String) o2);
    }

    @Override
    public String serialize(Object obj) {
        return (String) obj;
    }

    @Override
    public Object deserialize(String s) {
        return s;
    }
}
